### ISSUE TITLE
[Feature] Support Confidential Compute workloads using Intel SGX enclaves on CBL-Mariner as the AKS container host operating system

### ISSUE THREAD

**Issue by sdx-jkataja** [Azure/AKS#3815]
**Is your feature request related to a problem? Please describe.**
We have a Confidential Computing workload that uses the Intel Software Guard Extensions (SGX). We run the workload on AKS clusters with a node pool backed by Confidential Compute 'DC' type VMs.  We use the [Confidential Computing add-on for AKS](https://learn.microsoft.com/en-us/azure/confidential-computing/confidential-nodes-aks-addon)  to add the Intel SGX driver to the Linux kernel running on the AKS container host. This is currently available only for the Ubuntu version 18.04. This operating system has a large footprint and is already dated. We are evaluating CBL-Mariner as the base operating system image for AKS cluster nodes.  This Confidential Computing add-on for AKS does not currently appear to support this AKS container host operating system. The lack of Confidential Computing support blocks us from changing to CBL-Mariner.

**Describe the solution you'd like**
Ideally, the CBL-Mariner AKS container host image contains Intel SGX device driver out of the box, node initialization detects the DC type VM and inserts the kernel module automatically. The SGX devices are available to pods running on that node. I don't mind installing the [Intel Device Plugins Operator](https://intel.github.io/intel-device-plugins-for-kubernetes/cmd/operator/README.html) to provide the Kubernetes-layer support. This solution would not require the AKS add-on.

**Describe alternatives you've considered**
Alternatively, support CBL-Mariner in the Confidential Computing add-on for AKS.

**Additional context**
I found an article from 2022 that CBL-Mariner now allows building and installing kernel support for Intel SGX,. Even if this is possible to do, we would like to avoid creating and maintaining custom AKS container host images.

**Comment by sjwaight** [comment]
@jakjang can you please review and update status.

**Comment by jakjang** [comment]
Ack. @sdx-jkataja could you confirm if this is still an active ask? I can check in and follow up here afterwards.

**Comment by sdx-jkataja** [comment]
@jakjang Yes, as far as I know. I would start by checking if the build configuration for Intel SGX is enabled in the container host for AKS image.

**Comment by jakjang** [comment]
ack: apologies for the delayed response. I'll revert here once we have more details.

**Comment by jakjang** [comment]
Azure Linux currently does not support Intel SGX for their images. There is a hard limitation adding in support, but they are looking to move to TDX as an alternative. 

@sdx-jkataja Is there a reason Intel SGX is the choice here? Would Intel TDX be a viable alternative for your use case?

**Comment by sdx-jkataja** [comment]
Azure Linux appears to support Intel SGX to some degree:

- kernel configuration has the SGX option enabled: https://github.com/microsoft/azurelinux/blob/62084ef77ccbbdb2960466047ae16831f31ca0f0/SPECS/kernel/config#L481
- distribution contains the sgx-backwards-compatibility package for the changed device paths:
https://github.com/microsoft/azurelinux/tree/62084ef77ccbbdb2960466047ae16831f31ca0f0/SPECS/sgx-backwards-compatibility

My feature request was about extending this support to AKS Container Host images, so that Intel SGX would be supported with Azure Linux on AKS. This lack of support keeps us in old version of Ubuntu with a larger footprint for vulnerabilities.

@jakjang we deploy an application using Intel SGX on AKS. It cannot be ported to Intel TDX, which uses an entirely different model. So unfortunately this is not viable.

**Comment by jakjang** [comment]
Keeping this thread warm; thank you for the context! Interesting callout; `CONFIG_X86_SGX=y` does look like SGX should be available (or at least supported to some extent like you say). 

I'll investigate a little further with our OS team before getting back here with an update!

Edit: Chatting with some folks on our OS team; there's nothing technically infeasible around getting the feature enabled at the OS level (and as you pointed out, kernel even looks to have it enabled). However, there would be material work needed in the rest of the stack to actually get SGX working in Azure Linux. I'll add the ask on to their backlog.

**Comment by sdx-jkataja** [comment]
@jakjang thanks for your response. From my point of view there is appears to be no need to have the Confidential Computing add-on for AKS for Azure Linux. As long as we get OS support and run on the matching DC-family VMs, we can install the Node Feature Discovery operator and Intel Device Plugin operator on the cluster. This should cover all the layers.

**Comment by jakjang** [comment]
@sdx-jkataja Thanks for the engagement here; ask does makes sense. I've put the ask as on our Azure Linux team's backlog, but could I also trouble you to open up a feature request in the [Azure Linux GitHub](https://github.com/microsoft/azurelinux/issues) for specifically the OS support for SGX you mentioned in your latest comment?

**Comment by sdx-jkataja** [comment]
@jakjang  I don't have visibility to what else the AKS addon "confcom" contains in order to implement the Intel SGX support.  Therefore, I do not know what is missing and what to ask from Azure Linux.  Kernel support seems to have been added in the pull request: https://github.com/microsoft/azurelinux/pull/2154

According to AKS documentation the Cofidential Computing addon is required, but it is only available for Ubuntu 18. Perhaps all that is needed for it to work is enabling the driver in kernel used by AKS Container Host image. Even better if the driver is already included. But is this a supported configuration?

**Comment by jakjang** [comment]
> [@jakjang](https://github.com/jakjang) I don't have visibility to what else the AKS addon "confcom" contains in order to implement the Intel SGX support. Therefore, I do not know what is missing and what to ask from Azure Linux. Kernel support seems to have been added in the pull request: [microsoft/azurelinux#2154](https://github.com/microsoft/azurelinux/pull/2154)
> 
> According to AKS documentation the Cofidential Computing addon is required, but it is only available for Ubuntu 18. Perhaps all that is needed for it to work is enabling the driver in kernel used by AKS Container Host image. Even better if the driver is already included. But is this a supported configuration?

Sorry, definitely could've been clearer on my end. Ask to Azure Linux would purely be to bring up all the parts needed to support SGX in Azure Linux, pertaining to the comment above where as long as the OS support is there (and can be run on the proper VM sizes), it seems there'd be a path to get it working (installing device operator, etc.). 

To the point of the PR, while the kernel support was added, there are additional packages/work that would be needed to ensure that SGX capabilities are up and fully validated in Azure Linux; ask to Azure Linux would hit these aspects (quite broadly, "getting SGX properly enabled on Azure Linux" should cover this work).

To your point, right now the CC addon is enabled on for Ubuntu, but to get it working, OS pieces beyond just the Azure Linux kernel would be needed. Material work would be needed on the OS portion of the stack first.

**Comment by sdx-jkataja** [comment]
@jakjang it appears that the current Azure Linux Container Host image has the kernel option enabled:
```
$ uname -a
Linux retracted 6.6.92.2-2.azl3 #1 SMP PREEMPT_DYNAMIC Wed Jul  2 02:54:28 UTC 2025 x86_64 Linux
$ gunzip -d < /proc/config.gz | grep SGX
CONFIG_X86_SGX=y
CONFIG_X86_SGX_KVM=y
```
I have not yet tested the application to see if it works.
